### PR TITLE
Enable MSVC EH for Win64

### PR DESCRIPTION
**Pull request by kinke**
Enable @rainers #1168 for Win64 too and add x86 AppVeyor jobs.
Requires https://github.com/ldc-developers/druntime/pull/65 and a post-3.8 LLVM (Rainer's work has uncovered a few LLVM bugs whose near-instant fixes didn't make it into 3.8 final).

### CODE REVIEW

**Comment by kinke** [comment]
What AppVeyor doesn't show is that dmd-testsuite (debug+release) finally passes for both 32bit and 64bit MSVC targets! I.e. for the very first time for LDC on Windows in case you're wondering.
Awesome work Rainer, thanks again.

**Comment by smolt** [comment]
I second that :+1:

**Comment by rainers** [comment]
@kinke Thanks for pushing this.

> And thanks to @majnemer and all the other LLVM guys

I second that, too.

**Comment by kinke** [comment]
To recap: the only issues leading to test failures on Windows after this are:
* https://github.com/D-Programming-Language/phobos/pull/4034 (32-bit only)
* https://github.com/D-Programming-Language/phobos/pull/3976 (32-bit release only)
* LDC IR test issue https://github.com/ldc-developers/ldc/issues/1265 (32-bit only due to special mangling and calling convention)
* A Win64-specific fibers unittest in `core.thread` not working with inlining (a problem of the test itself).

**Comment by redstar** [comment]
Very well done! @kinke I think you deserve the honor to push the merge button. :-)

**Comment by kinke** [comment]
> I think you deserve the honor to push the merge button.

Let me first get the Debug x86 job green (with some cheating - by disabling the lit tests for now due to #1356).
The honour would be all @rainers'. We'll just have to get him to join the LDC devs group. :)

**Comment by kinke** [comment]
dmd-testsuite is now enabled too for the AppVeyor jobs (it doesn't take that much longer). There are some minor path-related issues I don't have on my local system, probably due to a different set of GNU utils (I assume), e.g.:
```
... compilable/test9680.sh
Error: Could not create output directory 'C:C:/Program Files/Git/projects/ninja-ldc/dmd-testsuite-debug/compilable'.
```
Otherwise, there are only issues wrt. Dan's new `runnable\ldc_cabi1.d` (on my system too - I've previously used a dmd-testsuite from a couple of weeks ago). The x64 job hits LLVM assertions - apparently it doesn't like the `align(16)` attribute:
```
Wrong types for attribute: byval inalloca nest noalias nocapture nonnull readnone readonly signext sret zeroext dereferenceable(1) dereferenceable_or_null(1)
  call void (i32, ...) @cvfun(i32 3, %ldc_cabi1.F4 noalias nocapture align 16 %509), !dbg !913
Wrong types for attribute: byval inalloca nest noalias nocapture nonnull readnone readonly signext sret zeroext dereferenceable(1) dereferenceable_or_null(1)
  call void (i32, ...) @cvfun(i32 4, %ldc_cabi1.D4 noalias nocapture align 16 %536), !dbg !918
Wrong types for attribute: byval inalloca nest noalias nocapture nonnull readnone readonly signext sret zeroext dereferenceable(1) dereferenceable_or_null(1)
  call void (i32, ...) @cvfun(i32 5, %ldc_cabi1.DHFA2 noalias nocapture align 16 %552), !dbg !921
Wrong types for attribute: byval inalloca nest noalias nocapture nonnull readnone readonly signext sret zeroext dereferenceable(1) dereferenceable_or_null(1)
  call void (i32, ...) @cvfun(i32 6, %ldc_cabi1.DHFA2a noalias nocapture align 16 %568), !dbg !924
LLVM ERROR: Broken function found, compilation aborted!
```
The x86 job uncovers mangling issues:
```
ldc_cabi2.cpp.obj : error LNK2019: unresolved external symbol "signed char a" (?a@@3CA) referenced in function _cretb1
ldc_cabi2.cpp.obj : error LNK2019: unresolved external symbol "signed char b" (?b@@3CA) referenced in function _cretb2
ldc_cabi2.cpp.obj : error LNK2019: unresolved external symbol "signed char c" (?c@@3CA) referenced in function _cretf4
ldc_cabi2.cpp.obj : error LNK2019: unresolved external symbol "signed char d" (?d@@3CA) referenced in function _cretf4
ldc_cabi2.cpp.obj : error LNK2019: unresolved external symbol "unsigned int errors" (?errors@@3IA) referenced in function _cretb1
```

**Comment by smolt** [comment]
The mangling error I think is a flaw in ldc_cabi2.cpp.  Those vars should be extern "C".

**Comment by kinke** [comment]
> Those vars should be extern "C".

That's right, I'll fix them, thanks for the hint. Now we hit some test assertions for x86.
The (current) Win64 issue only applies to variadic functions.

**Comment by kinke** [comment]
dmd-testsuite passes again on Win64, both on my box and for the AppVeyor x64 Debug job.

**Comment by kinke** [comment]
The Win32 `ldc_cabi1` failures are also vararg related. Fix is on its way.

**Comment by kinke** [comment]
Alright, now all tests are enabled for AppVeyor, but lit test failures are ignored for x86 jobs because of failing `align.d` due to issue #1356.

The debug jobs are green. The release ones fail (and are allowed to fail)
* on Win64 only because of the inline-incompatible `core.thread` unittest (Win64 specific test, hence no problem for all other platforms), and
* on Win32 only due to @smolt's https://github.com/D-Programming-Language/phobos/pull/3976 not having been merged yet.

This really elevates Windows MSVC to a first class target for LDC. As I mentioned earlier, it depends on a post-3.8 LLVM. @rainers even included a TLS alignment bugfix for Windows < 8.1. The recent addition of alignment attributes for IR arguments has uncovered quite a few alignment issues due to the optimizer apparently making good use of that information, so that's a good sign too.

**Comment by smolt** [comment]
This is awesome!

File: appveyor.yml [hunk 1]
```diff
@@ -46,44 +45,65 @@ install:
...
   - ps: Start-FileDownload 'http://downloads.dlang.org/releases/2015/dmd.2.069.2.windows.7z' -FileName 'dmd2.7z'
   - 7z x dmd2.7z > nul
-  - set PATH=%PATH%;c:\projects\dmd2\windows\bin
-  - set DC=c:\projects\dmd2\windows\bin\dmd.exe
-  - dmd --version
+  - set DMD=%CD%\dmd2\windows\bin\dmd.exe
+  - dmd2\windows\bin\dmd.exe --version
+  # Download & extract GNU make + utils (for dmd-testsuite)
+  - bash --version
+  - ps: Start-FileDownload 'https://dl.dropboxusercontent.com/s/d0cqmxf9arbjn27/make-3.81.7z?dl=0' -FileName 'make.7z'
```
Comment (CyberShadow) [inline appveyor.yml]: Hey, just wanted to ask, what's the original source of this file?
(Looking into AppVeyor for Digger...)

Comment (kinke) [inline appveyor.yml]: Yes, but it contains some additional utils. See http://wiki.dlang.org/Building_and_hacking_LDC_on_Windows_using_MSVC#Running_the_dmd-testsuite_tests.

**Pull request merged**